### PR TITLE
#95 - validate filename on upload

### LINKED ISSUE #95: Return 400 if file name does not start with project name
Python simple repository returns 400 if uploaded file name does not start with project name specified in `PKG_INFO` metadata:

```bash
MacBook-Pro:~ alena$ python3 -m twine upload --repository-url https://test.pypi.org/legacy/ -u olena -p '!@$FZqTSG6iEQeX' --verbose ./PycharmProjects/sampleproject/dist/*
Uploading distributions to https://test.pypi.org/legacy/
  ./PycharmProjects/sampleproject/dist/one.whl (3.0 KB)
Uploading one.whl
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5.58k/5.58k [00:01<00:00, 4.75kB/s]
Content received from server:
<html>
 <head>
  <title>400 Start filename for 'olena-sample' with 'olena-sample'.</title>
 </head>
 <body>
  <h1>400 Start filename for 'olena-sample' with 'olena-sample'.</h1>
  The server could not comply with the request since it is either malformed or otherwise incorrect.<br/><br/>
Start filename for &#x27;olena-sample&#x27; with &#x27;olena-sample&#x27;.


 </body>
</html>
HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
Start filename for 'olena-sample' with 'olena-sample'.
```
Project name in the example above is `olena-sample`.

If uploaded artifact is not properly named, then it's not possible to install it. So, I think we should check file name correspondence with metadata and return 400 if smth is wrong.

### PR DESCRIPTION
**Pull request by olenagerasimova**
Added filename validation on upload to close #95.

### CODE REVIEW

File: src/main/java/com/artipie/pypi/meta/ValidFilename.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        this.filename = filename;
+    }
+
+    /**
+     * Is filename valid?
+     * @return True if filename corresponds to project metadata, false - otherwise.
+     */
+    public boolean valid() {
+        final Matcher matcher;
+        if (this.filename.endsWith("whl")) {
```
Comment (olegmoz) [inline src/main/java/com/artipie/pypi/meta/ValidFilename.java:76]: @olenagerasimova this check for end of string actually duplicates what we have in `WHEEL_PTRN`. I'd suggest to test `this.filename` against stream of patterns and take first one that matches

File: src/test/java/com/artipie/pypi/http/WheelSliceTest.java [hunk 2]
```diff
@@ -106,6 +106,35 @@ void savesContentByNormalizedNameAndReturnsOk() throws IOException {
...
+                        )
+                    )
+                )
+            ),
+            new RsHasStatus(RsStatus.BAD_REQUEST)
+        );
+        MatcherAssert.assertThat(
+            "Storage is empty",
+            storage.list(Key.ROOT).join().size(),
+            new IsEqual<>(0)
```
Comment (olegmoz) [inline src/test/java/com/artipie/pypi/http/WheelSliceTest.java:134]: @olenagerasimova you may check that collection is empty using `IsEmptyCollection` matcher. This way mismatch description is going to be better and code is a little bit more simple.

File: src/main/java/com/artipie/pypi/meta/ValidFilename.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+import com.artipie.pypi.NormalizedProjectName;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Python package valid filename.
+ * @since 0.6
+ */
+public final class ValidFilename {
```
[the remark this answers is not in the document]
Comment (olenagerasimova) [reply src/main/java/com/artipie/pypi/meta/ValidFilename.java:34]: @olegmoz I'm planning to remove `NormalizedProjectName.FromFilename` (see #104) as it is not used anymore and `NormalizedProjectName.FromFilename` purpose was to obtain/normalize project name from filename when this class purpose is to validate filename.

**Comment by olenagerasimova** [comment]
@olegmoz thanks, corrected and answered, take a look please

**Comment by 0crat** [comment]
Job `gh:artipie/pypi-adapter#102` is not assigned, can't get performer

**Pull request merged**